### PR TITLE
Don't return secrets in the API & Only update credentials when requested

### PR DESCRIPTION
**Pull request by sherifnada**
## What
1. Adds `JsonSecretsProcessor` which exposes methods for expunging or absorbing secrets to and from configuration objects. 
1. Using `JsonSecretsProcessor`, change Source/Destination endpoints to never return any sensitive configs. 
1. Add a `shouldUpdateCredentials` flag to `SourceUpdate`/`DestinationUpdate` to indicate whether the values of sensitive fields should be updated or not. If not, ignores the value of secrets even if they are set in the input to the update method. 

## Reading Order
1. `JsonSecretsProcessor.java` + `JsonSecretsProcessorTest.java`
1. `config.yaml`
1. `SourceHandler.java`
1. `DestinationHandler.java`
1. anything else

## TODO
- [x] fix tests

### CODE REVIEW

**Comment by sherifnada** [comment]
Publishing this for review to get feedback since the approach is slightly different than what was laid out in #986 in that we do not introspect into the value of secrets to figure out if they should be updated or not.

File: airbyte-api/src/main/openapi/config.yaml [hunk 1]
```diff
@@ -1135,6 +1135,8 @@ components:
           $ref: "#/components/schemas/SourceId"
         connectionConfiguration:
           $ref: "#/components/schemas/SourceConfiguration"
+        updateConfigurationSecrets:
```
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:1138]: this strategy is different from the originally proposed one in the issue https://github.com/airbytehq/airbyte/issues/986. Why the change?

How does this implementation handle the case where a configuration has 2 secrets and only one is updated?

File: airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+   * @param schema Schema containing secret annotations
+   * @param obj    Object containing potentially secret fields
+   * @return
+   */
+  public JsonNode removeSecrets(JsonNode obj, JsonNode schema) {
+    assertValidSchema(schema);
+    Preconditions.checkArgument(schema.isObject());
+
+    ObjectNode properties = (ObjectNode) schema.get(PROPERTIES_FIELD);
+    JsonNode copy = obj.deepCopy();
```
Comment (cgardens) [inline airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java:52]: `Jsons.clone` is an option too. I don't know what guarantees `deepCopy` gives but if you do, then good to go!

File: airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    return dstCopy;
+  }
+
+  private static boolean isSecret(JsonNode obj) {
+    return obj.isObject() && obj.has(AIRBYTE_SECRET_FIELD) && obj.get(AIRBYTE_SECRET_FIELD).asBoolean();
+  }
+
+  private static void assertValidSchema(JsonNode node) {
```
Comment (cgardens) [inline airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java:95]: why do we want to throw is a schema doesn't match our expectations? don't we just want to do nothing? e.g. if there is no top level properties and it is just a oneOf we don't want to fail do we?

**Review by cgardens — CHANGES_REQUESTED** [review]
Seems like you went for a different protocol than the one described in the issue, would like to understand the motivations for that change before reviewing more! https://github.com/airbytehq/airbyte/issues/986

**Comment by sherifnada** [comment]
Addressing comments, un-requested review for now

**Comment by sherifnada** [comment]
Ready for another round of review

**Review by cgardens — APPROVED** [review]
lgtm! thanks.

File: airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java (hunk 2 above)
Comment (sherifnada) [reply airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java:58]: deepCopy can return a different type e.g: `ObjectNode`. Just a small thing. I didn't realize jsons.clone was a thing

File: airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java (hunk 3 above)
Comment (sherifnada) [reply airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSecretsProcessor.java:105]: fair enough. will fix

**Review by michel-tricot — COMMENTED** [review]
🎉 !!
